File: src/l1mergeintervals/CS/MergeIntervals.Algo/IntervalMerger.cs
using System.Collections.Generic;

namespace MergeIntervals.Algo;

public static class IntervalMerger
{
    public static List<int[]> MergeIntervals(List<int[]> intervals)
    {
        if (intervals.Count == 0)
        {
            return new List<int[]>();
        }

        var result = new List<int[]>
        {
            intervals[0]
        };

        for (int i = 1; i < intervals.Count; i++)
        {
            var lastAddedInterval = result[^1];
            var currentInterval = intervals[i];
            var curStart = currentInterval[0];
            var curEnd = currentInterval[1];

            if (lastAddedInterval[1] >= curStart)
            {
                lastAddedInterval[1] = System.Math.Max(lastAddedInterval[1], curEnd);
            }
            else
            {
                result.Add(currentInterval);
            }
        }

        return result;
    }
}

File: src/l1mergeintervals/Java/com/example/mergeintervals/IntervalMerger.java
package com.example.mergeintervals;

import java.util.ArrayList;
import java.util.List;

public class IntervalMerger {
    public static int[][] mergeIntervals(int[][] intervals) {
        if (intervals.length == 0) {
            return new int[0][];
        }

        List<int[]> result = new ArrayList<>();
        result.add(intervals[0]);

        for (int i = 1; i < intervals.length; i++) {
            int[] lastAddedInterval = result.get(result.size() - 1);
            int[] currentInterval = intervals[i];
            int curStart = currentInterval[0];
            int curEnd = currentInterval[1];

            if (lastAddedInterval[1] >= curStart) {
                lastAddedInterval[1] = Math.max(lastAddedInterval[1], curEnd);
            } else {
                result.add(currentInterval);
            }
        }

        return result.toArray(new int[result.size()][]);
    }
}
